[Bug] [Zeta] NPE in CheckpointCoordinator.completePendingCheckpoint()

### Search before asking

- [x] I had searched in the [issues](https://github.com/apache/seatunnel/issues?q=is%3Aissue+label%3A%22bug%22) and found no similar issues.


### What happened

When `notifyCompleted()` encounters an exception internally, it calls `handleCoordinatorError()` which in turn calls `cleanPendingCheckpoint()`, which clears pendingCheckpoints. The subsequent `pendingCheckpoints.remove(checkpointId)` then returns null, and the original chained call `abortCheckpointTimeoutFutureWhenIsCompleted()` would throw a NullPointerException.

### SeaTunnel Version

2.3.13

### SeaTunnel Config

```conf
None
```

### Running Command

```shell
None
```

### Error Exception

```log
Caused by: java.lang.NullPointerException
	at org.apache.seatunnel.engine.server.checkpoint.CheckpointCoordinator.completePendingCheckpoint(CheckpointCoordinator.java:1004)
	at org.apache.seatunnel.engine.server.checkpoint.CheckpointCoordinator.completePendingCheckpoint(CheckpointCoordinator.java:950)
	... 66 more
```

### Zeta or Flink or Spark Version

Zeta

### Java or Scala Version

_No response_

### Screenshots

<img width="1072" height="455" alt="Image" src="https://github.com/user-attachments/assets/27e71974-cb1d-4567-9cd3-7191e0801773" />

### Are you willing to submit PR?

- [x] Yes I am willing to submit a PR!

### Code of Conduct

- [x] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

--- a/seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/checkpoint/CheckpointCoordinator.java
+++ b/seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/checkpoint/CheckpointCoordinator.java
@@ -1001,7 +1001,10 @@ public class CheckpointCoordinator {
             checkpointMonitorService.onCheckpointCompleted(completedCheckpoint, stateSize);
         }
         notifyCompleted(completedCheckpoint);
-        pendingCheckpoints.remove(checkpointId).abortCheckpointTimeoutFutureWhenIsCompleted();
+        PendingCheckpoint pendingCheckpoint = pendingCheckpoints.remove(checkpointId);
+        if (pendingCheckpoint != null) {
+            pendingCheckpoint.abortCheckpointTimeoutFutureWhenIsCompleted();
+        }
         pendingCounter.decrementAndGet();
 
         if (isCompleted()) {

--- a/seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/checkpoint/CheckpointCoordinator.java
+++ b/seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/checkpoint/CheckpointCoordinator.java
@@ -358,7 +358,9 @@ public class CheckpointCoordinator {
         }
         InvocationFuture<?>[] futures = notifyTaskStart();
         CompletableFuture.allOf(futures).join();
-        notifyCompleted(latestCompletedCheckpoint);
+        if (!notifyCompleted(latestCompletedCheckpoint)) {
+            return;
+        }
         if (coordinatorConfig.isCheckpointEnable()) {
             LOG.info("checkpoint is enabled, start schedule trigger pending checkpoint.");
             scheduleTriggerPendingCheckpoint(coordinatorConfig.getCheckpointInterval());
@@ -369,7 +371,7 @@ public class CheckpointCoordinator {
     }
 
     @VisibleForTesting
-    protected void notifyCompleted(CompletedCheckpoint completedCheckpoint) {
+    protected boolean notifyCompleted(CompletedCheckpoint completedCheckpoint) {
         if (completedCheckpoint != null) {
             try {
                 LOG.info(
@@ -390,8 +392,10 @@ public class CheckpointCoordinator {
                         "notify checkpoint completed failed",
                         e,
                         CheckpointCloseReason.CHECKPOINT_NOTIFY_COMPLETE_FAILED);
+                return false;
             }
         }
+        return true;
     }
 
     public InvocationFuture<?>[] notifyTaskStart() {
@@ -490,7 +494,9 @@ public class CheckpointCoordinator {
         shutdown = false;
         if (alreadyStarted) {
             isAllTaskReady.set(true);
-            notifyCompleted(latestCompletedCheckpoint);
+            if (!notifyCompleted(latestCompletedCheckpoint)) {
+                return;
+            }
             tryTriggerPendingCheckpoint(CHECKPOINT_TYPE);
         } else {
             isAllTaskReady.set(false);
@@ -1000,7 +1006,9 @@ public class CheckpointCoordinator {
             long stateSize = CheckpointMonitorService.calculateStateSize(completedCheckpoint);
             checkpointMonitorService.onCheckpointCompleted(completedCheckpoint, stateSize);
         }
-        notifyCompleted(completedCheckpoint);
+        if (!notifyCompleted(completedCheckpoint)) {
+            return;
+        }
         PendingCheckpoint pendingCheckpoint = pendingCheckpoints.remove(checkpointId);
         if (pendingCheckpoint != null) {
             pendingCheckpoint.abortCheckpointTimeoutFutureWhenIsCompleted();

--- a/seatunnel-engine/seatunnel-engine-server/src/test/java/org/apache/seatunnel/engine/server/checkpoint/CheckpointCoordinatorTest.java
+++ b/seatunnel-engine/seatunnel-engine-server/src/test/java/org/apache/seatunnel/engine/server/checkpoint/CheckpointCoordinatorTest.java
@@ -51,6 +51,7 @@ import java.util.HashSet;
 import java.util.List;
 import java.util.Map;
 import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.ExecutionException;
 import java.util.concurrent.ExecutorService;
 import java.util.concurrent.Executors;
@@ -370,6 +371,108 @@ public class CheckpointCoordinatorTest
         Assertions.assertEquals(1, checkpointManager.operations.size());
 
         executor.shutdownNow();
+    }
+
+    /**
+     * Regression test for the NPE bug in {@code completePendingCheckpoint}.
+     *
+     * <p>When {@code notifyCompleted()} encounters an exception internally, it calls {@code
+     * handleCoordinatorError} which in turn calls {@code cleanPendingCheckpoint}, which clears
+     * {@code pendingCheckpoints}. The subsequent {@code pendingCheckpoints.remove(checkpointId)}
+     * then returns {@code null}, and the original chained call {@code
+     * .abortCheckpointTimeoutFutureWhenIsCompleted()} would throw a {@link NullPointerException}.
+     *
+     * <p>After the fix, {@code completePendingCheckpoint} performs a null-check before calling
+     * {@code abortCheckpointTimeoutFutureWhenIsCompleted()}, so no NPE is thrown.
+     */
+    @Test
+    void testCompletePendingCheckpointShouldNotThrowNPEWhenNotifyCompletedClearsPendingMap() {
+        CheckpointConfig checkpointConfig = new CheckpointConfig();
+        checkpointConfig.setStorage(new CheckpointStorageConfig());
+
+        TaskLocation taskLocation = new TaskLocation(new TaskGroupLocation(1L, 1, 1), 1, 1);
+        Map<Integer, CheckpointPlan> planMap = new HashMap<>();
+        planMap.put(
+                1,
+                CheckpointPlan.builder()
+                        .pipelineId(1)
+                        .pipelineSubtasks(Collections.singleton(taskLocation))
+                        .startingSubtasks(Collections.singleton(taskLocation))
+                        .build());
+
+        ExecutorService executorService = Executors.newCachedThreadPool();
+        try {
+            CheckpointManager checkpointManager =
+                    new CheckpointManager(
+                            1L,
+                            false,
+                            nodeEngine,
+                            null,
+                            planMap,
+                            checkpointConfig,
+                            server.getCheckpointService().getCheckpointStorage(),
+                            executorService,
+                            nodeEngine.getHazelcastInstance().getMap(IMAP_RUNNING_JOB_STATE),
+                            null);
+
+            CheckpointCoordinator coordinator = checkpointManager.getCheckpointCoordinator(1);
+
+            CheckpointCoordinator spyCoordinator = Mockito.spy(coordinator);
+            Mockito.doAnswer(
+                            invocation -> {
+                                @SuppressWarnings("unchecked")
+                                ConcurrentHashMap<Long, PendingCheckpoint> map =
+                                        (ConcurrentHashMap<Long, PendingCheckpoint>)
+                                                ReflectionUtils.getField(
+                                                                spyCoordinator,
+                                                                "pendingCheckpoints")
+                                                        .orElse(null);
+                                if (map != null) {
+                                    map.clear();
+                                }
+                                return null;
+                            })
+                    .when(spyCoordinator)
+                    .notifyCompleted(Mockito.any());
+
+            long checkpointId = 1L;
+            CompletedCheckpoint completedCheckpoint =
+                    new CompletedCheckpoint(
+                            1L,
+                            1,
+                            checkpointId,
+                            System.currentTimeMillis(),
+                            CheckpointType.CHECKPOINT_TYPE,
+                            System.currentTimeMillis(),
+                            new HashMap<>(),
+                            new HashMap<>());
+
+            PendingCheckpoint pendingCheckpoint =
+                    new PendingCheckpoint(
+                            1L,
+                            1,
+                            checkpointId,
+                            System.currentTimeMillis(),
+                            CheckpointType.CHECKPOINT_TYPE,
+                            new HashSet<>(),
+                            new HashMap<>(),
+                            new HashMap<>());
+            ConcurrentHashMap<Long, PendingCheckpoint> pendingCheckpoints =
+                    (ConcurrentHashMap<Long, PendingCheckpoint>)
+                            ReflectionUtils.getField(spyCoordinator, "pendingCheckpoints")
+                                    .orElseThrow(
+                                            () ->
+                                                    new IllegalStateException(
+                                                            "pendingCheckpoints field not found"));
+            pendingCheckpoints.put(checkpointId, pendingCheckpoint);
+
+            Assertions.assertDoesNotThrow(
+                    () -> spyCoordinator.completePendingCheckpoint(completedCheckpoint),
+                    "completePendingCheckpoint must not throw NullPointerException when "
+                            + "pendingCheckpoints is cleared by notifyCompleted");
+        } finally {
+            executorService.shutdownNow();
+        }
     }
 }
 

--- a/seatunnel-engine/seatunnel-engine-server/src/test/java/org/apache/seatunnel/engine/server/checkpoint/CheckpointCoordinatorTest.java
+++ b/seatunnel-engine/seatunnel-engine-server/src/test/java/org/apache/seatunnel/engine/server/checkpoint/CheckpointCoordinatorTest.java
@@ -34,6 +34,8 @@ import org.apache.seatunnel.engine.server.task.statemachine.SeaTunnelTaskState;
 
 import org.junit.jupiter.api.Assertions;
 import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.condition.DisabledOnOs;
+import org.junit.jupiter.api.condition.OS;
 import org.mockito.MockedStatic;
 import org.mockito.Mockito;
 
@@ -386,6 +388,7 @@ public class CheckpointCoordinatorTest
      * {@code abortCheckpointTimeoutFutureWhenIsCompleted()}, so no NPE is thrown.
      */
     @Test
+    @DisabledOnOs(OS.WINDOWS)
     void testCompletePendingCheckpointShouldNotThrowNPEWhenNotifyCompletedClearsPendingMap() {
         CheckpointConfig checkpointConfig = new CheckpointConfig();
         checkpointConfig.setStorage(new CheckpointStorageConfig());

--- a/seatunnel-engine/seatunnel-engine-server/src/test/java/org/apache/seatunnel/engine/server/checkpoint/CheckpointCoordinatorTest.java
+++ b/seatunnel-engine/seatunnel-engine-server/src/test/java/org/apache/seatunnel/engine/server/checkpoint/CheckpointCoordinatorTest.java
@@ -417,8 +417,8 @@ public class CheckpointCoordinatorTest
 
     /**
      * Regression: when {@code notifyCompleted()} fails (returns {@code false}), {@code
-     * completePendingCheckpoint} must return immediately without decrementing {@code pendingCounter}
-     * or executing any other "success path" logic.
+     * completePendingCheckpoint} must return immediately without decrementing {@code
+     * pendingCounter} or executing any other "success path" logic.
      *
      * <p>Before the fix the chained call on the {@code null} result of {@code
      * pendingCheckpoints.remove()} caused a {@link NullPointerException}.
@@ -437,8 +437,7 @@ public class CheckpointCoordinatorTest
                                 @SuppressWarnings("unchecked")
                                 ConcurrentHashMap<Long, PendingCheckpoint> map =
                                         (ConcurrentHashMap<Long, PendingCheckpoint>)
-                                                ReflectionUtils.getField(
-                                                                spy, "pendingCheckpoints")
+                                                ReflectionUtils.getField(spy, "pendingCheckpoints")
                                                         .orElse(null);
                                 if (map != null) {
                                     map.clear();
@@ -552,8 +551,8 @@ public class CheckpointCoordinatorTest
     }
 
     /**
-     * Regression: when {@code notifyCompleted()} fails inside {@code restoreCoordinator(true)},
-     * the method must return immediately and must NOT call {@code tryTriggerPendingCheckpoint}.
+     * Regression: when {@code notifyCompleted()} fails inside {@code restoreCoordinator(true)}, the
+     * method must return immediately and must NOT call {@code tryTriggerPendingCheckpoint}.
      */
     @Test
     void testRestoreCoordinatorShouldNotTriggerCheckpointWhenNotifyCompletedFails() {

--- a/seatunnel-engine/seatunnel-engine-server/src/test/java/org/apache/seatunnel/engine/server/checkpoint/CheckpointCoordinatorTest.java
+++ b/seatunnel-engine/seatunnel-engine-server/src/test/java/org/apache/seatunnel/engine/server/checkpoint/CheckpointCoordinatorTest.java
@@ -22,6 +22,7 @@ import org.apache.seatunnel.engine.checkpoint.storage.api.CheckpointStorage;
 import org.apache.seatunnel.engine.common.config.server.CheckpointConfig;
 import org.apache.seatunnel.engine.common.config.server.CheckpointStorageConfig;
 import org.apache.seatunnel.engine.common.utils.concurrent.CompletableFuture;
+import org.apache.seatunnel.engine.core.checkpoint.CheckpointIDCounter;
 import org.apache.seatunnel.engine.core.checkpoint.CheckpointType;
 import org.apache.seatunnel.engine.server.AbstractSeaTunnelServerTest;
 import org.apache.seatunnel.engine.server.checkpoint.monitor.CheckpointMonitorService;
@@ -34,8 +35,6 @@ import org.apache.seatunnel.engine.server.task.statemachine.SeaTunnelTaskState;
 
 import org.junit.jupiter.api.Assertions;
 import org.junit.jupiter.api.Test;
-import org.junit.jupiter.api.condition.DisabledOnOs;
-import org.junit.jupiter.api.condition.OS;
 import org.mockito.MockedStatic;
 import org.mockito.Mockito;
 
@@ -59,6 +58,7 @@ import java.util.concurrent.ExecutorService;
 import java.util.concurrent.Executors;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicInteger;
 import java.util.concurrent.atomic.AtomicLong;
 
 import static org.apache.seatunnel.engine.common.Constant.IMAP_RUNNING_JOB_STATE;
@@ -375,67 +375,77 @@ public class CheckpointCoordinatorTest
         executor.shutdownNow();
     }
 
+    // ------------------------------------------------------------------
+    // Regression tests: notifyCompleted returns false → callers bail out
+    // ------------------------------------------------------------------
+
     /**
-     * Regression test for the NPE bug in {@code completePendingCheckpoint}.
-     *
-     * <p>When {@code notifyCompleted()} encounters an exception internally, it calls {@code
-     * handleCoordinatorError} which in turn calls {@code cleanPendingCheckpoint}, which clears
-     * {@code pendingCheckpoints}. The subsequent {@code pendingCheckpoints.remove(checkpointId)}
-     * then returns {@code null}, and the original chained call {@code
-     * .abortCheckpointTimeoutFutureWhenIsCompleted()} would throw a {@link NullPointerException}.
-     *
-     * <p>After the fix, {@code completePendingCheckpoint} performs a null-check before calling
-     * {@code abortCheckpointTimeoutFutureWhenIsCompleted()}, so no NPE is thrown.
+     * Helper: build a minimal {@link CheckpointCoordinator} whose external dependencies are all
+     * mocked, so the test never touches Hazelcast / Hadoop I/O.
      */
-    @Test
-    @DisabledOnOs(OS.WINDOWS)
-    void testCompletePendingCheckpointShouldNotThrowNPEWhenNotifyCompletedClearsPendingMap() {
+    private CheckpointCoordinator buildMinimalCoordinator(ExecutorService executorService) {
         CheckpointConfig checkpointConfig = new CheckpointConfig();
         checkpointConfig.setStorage(new CheckpointStorageConfig());
 
         TaskLocation taskLocation = new TaskLocation(new TaskGroupLocation(1L, 1, 1), 1, 1);
-        Map<Integer, CheckpointPlan> planMap = new HashMap<>();
-        planMap.put(
-                1,
+        CheckpointPlan plan =
                 CheckpointPlan.builder()
                         .pipelineId(1)
                         .pipelineSubtasks(Collections.singleton(taskLocation))
                         .startingSubtasks(Collections.singleton(taskLocation))
-                        .build());
+                        .build();
 
+        CheckpointManager mockManager = Mockito.mock(CheckpointManager.class);
+        CheckpointStorage mockStorage = Mockito.mock(CheckpointStorage.class);
+        CheckpointIDCounter mockIdCounter = Mockito.mock(CheckpointIDCounter.class);
+        @SuppressWarnings("unchecked")
+        IMap<Object, Object> mockIMap = Mockito.mock(IMap.class);
+
+        return new CheckpointCoordinator(
+                mockManager,
+                mockStorage,
+                checkpointConfig,
+                1L,
+                plan,
+                mockIdCounter,
+                null,
+                executorService,
+                mockIMap,
+                false,
+                null);
+    }
+
+    /**
+     * Regression: when {@code notifyCompleted()} fails (returns {@code false}), {@code
+     * completePendingCheckpoint} must return immediately without decrementing {@code pendingCounter}
+     * or executing any other "success path" logic.
+     *
+     * <p>Before the fix the chained call on the {@code null} result of {@code
+     * pendingCheckpoints.remove()} caused a {@link NullPointerException}.
+     */
+    @Test
+    void testCompletePendingCheckpointShouldReturnEarlyWhenNotifyCompletedFails() {
         ExecutorService executorService = Executors.newCachedThreadPool();
         try {
-            CheckpointManager checkpointManager =
-                    new CheckpointManager(
-                            1L,
-                            false,
-                            nodeEngine,
-                            null,
-                            planMap,
-                            checkpointConfig,
-                            server.getCheckpointService().getCheckpointStorage(),
-                            executorService,
-                            nodeEngine.getHazelcastInstance().getMap(IMAP_RUNNING_JOB_STATE),
-                            null);
+            CheckpointCoordinator coordinator = buildMinimalCoordinator(executorService);
+            CheckpointCoordinator spy = Mockito.spy(coordinator);
 
-            CheckpointCoordinator coordinator = checkpointManager.getCheckpointCoordinator(1);
-
-            CheckpointCoordinator spyCoordinator = Mockito.spy(coordinator);
+            // Mock notifyCompleted to simulate failure: clear pendingCheckpoints (as
+            // handleCoordinatorError would) and return false.
             Mockito.doAnswer(
                             invocation -> {
                                 @SuppressWarnings("unchecked")
                                 ConcurrentHashMap<Long, PendingCheckpoint> map =
                                         (ConcurrentHashMap<Long, PendingCheckpoint>)
                                                 ReflectionUtils.getField(
-                                                                spyCoordinator,
-                                                                "pendingCheckpoints")
+                                                                spy, "pendingCheckpoints")
                                                         .orElse(null);
                                 if (map != null) {
                                     map.clear();
                                 }
-                                return null;
+                                return false;
                             })
-                    .when(spyCoordinator)
+                    .when(spy)
                     .notifyCompleted(Mockito.any());
 
             long checkpointId = 1L;
@@ -460,19 +470,107 @@ public class CheckpointCoordinatorTest
                             new HashSet<>(),
                             new HashMap<>(),
                             new HashMap<>());
+
+            @SuppressWarnings("unchecked")
             ConcurrentHashMap<Long, PendingCheckpoint> pendingCheckpoints =
                     (ConcurrentHashMap<Long, PendingCheckpoint>)
-                            ReflectionUtils.getField(spyCoordinator, "pendingCheckpoints")
+                            ReflectionUtils.getField(spy, "pendingCheckpoints")
                                     .orElseThrow(
                                             () ->
                                                     new IllegalStateException(
                                                             "pendingCheckpoints field not found"));
             pendingCheckpoints.put(checkpointId, pendingCheckpoint);
 
+            // Set pendingCounter to 1 so we can verify it is NOT decremented after failure.
+            AtomicInteger pendingCounter =
+                    (AtomicInteger)
+                            ReflectionUtils.getField(spy, "pendingCounter")
+                                    .orElseThrow(
+                                            () ->
+                                                    new IllegalStateException(
+                                                            "pendingCounter field not found"));
+            pendingCounter.set(1);
+
+            // Must not throw, and must not execute the success path.
             Assertions.assertDoesNotThrow(
-                    () -> spyCoordinator.completePendingCheckpoint(completedCheckpoint),
-                    "completePendingCheckpoint must not throw NullPointerException when "
-                            + "pendingCheckpoints is cleared by notifyCompleted");
+                    () -> spy.completePendingCheckpoint(completedCheckpoint),
+                    "completePendingCheckpoint must not throw when notifyCompleted fails");
+
+            // pendingCounter must remain 1 – the success path (decrementAndGet) was skipped.
+            Assertions.assertEquals(
+                    1,
+                    pendingCounter.get(),
+                    "pendingCounter must not be decremented when notifyCompleted fails");
+        } finally {
+            executorService.shutdownNow();
+        }
+    }
+
+    /**
+     * Regression: when {@code notifyCompleted()} fails inside {@code allTaskReady()}, the method
+     * must return immediately and must NOT schedule the next checkpoint trigger.
+     */
+    @Test
+    void testAllTaskReadyShouldNotScheduleCheckpointWhenNotifyCompletedFails() {
+        ExecutorService executorService = Executors.newCachedThreadPool();
+        try {
+            CheckpointCoordinator coordinator = buildMinimalCoordinator(executorService);
+            CheckpointCoordinator spy = Mockito.spy(coordinator);
+
+            // notifyTaskStart() must return a non-null empty array so allOf(...).join() succeeds.
+            Mockito.doReturn(new com.hazelcast.spi.impl.operationservice.impl.InvocationFuture[0])
+                    .when(spy)
+                    .notifyTaskStart();
+
+            // Simulate notifyCompleted failure.
+            Mockito.doReturn(false).when(spy).notifyCompleted(Mockito.any());
+
+            // Set all tasks to READY_START so allTaskReady() passes the guard checks.
+            Map<Long, SeaTunnelTaskState> taskStatus = spy.getPipelineTaskStatus();
+            CheckpointPlan plan =
+                    (CheckpointPlan)
+                            ReflectionUtils.getField(spy, "plan")
+                                    .orElseThrow(
+                                            () ->
+                                                    new IllegalStateException(
+                                                            "plan field not found"));
+            plan.getPipelineSubtasks()
+                    .forEach(t -> taskStatus.put(t.getTaskID(), SeaTunnelTaskState.READY_START));
+
+            // Invoke allTaskReady() via the package-private reportedTask path by directly calling
+            // the protected restoreCoordinator and then manipulating state, or call allTaskReady
+            // via reflection since it is private.
+            ReflectionUtils.invoke(spy, "allTaskReady");
+
+            // scheduleTriggerPendingCheckpoint must NOT have been called.
+            Mockito.verify(spy, Mockito.never())
+                    .scheduleTriggerPendingCheckpoint(
+                            Mockito.any(CheckpointType.class), Mockito.anyLong());
+        } finally {
+            executorService.shutdownNow();
+        }
+    }
+
+    /**
+     * Regression: when {@code notifyCompleted()} fails inside {@code restoreCoordinator(true)},
+     * the method must return immediately and must NOT call {@code tryTriggerPendingCheckpoint}.
+     */
+    @Test
+    void testRestoreCoordinatorShouldNotTriggerCheckpointWhenNotifyCompletedFails() {
+        ExecutorService executorService = Executors.newCachedThreadPool();
+        try {
+            CheckpointCoordinator coordinator = buildMinimalCoordinator(executorService);
+            CheckpointCoordinator spy = Mockito.spy(coordinator);
+
+            // Simulate notifyCompleted failure.
+            Mockito.doReturn(false).when(spy).notifyCompleted(Mockito.any());
+
+            // alreadyStarted=true is the branch that calls notifyCompleted.
+            spy.restoreCoordinator(true);
+
+            // tryTriggerPendingCheckpoint must NOT have been called.
+            Mockito.verify(spy, Mockito.never())
+                    .tryTriggerPendingCheckpoint(Mockito.any(CheckpointType.class));
         } finally {
             executorService.shutdownNow();
         }

